test/common/memory/memory_release_test.cc

```diff
@@ -2,71 +2,143 @@
-  uint64_t words_[ArraySize];
-};
+static const int MB = 1048576;
 
 class MemoryReleaseTest : public testing::Test {
 protected:
-  void initialiseAllocator(const uint64_t background_release_rate) {
-    memory_allocator_ = std::make_unique<Memory::Allocator>(Thread::threadFactoryForTest(),
-                                                            background_release_rate);
+  MemoryReleaseTest()
+      : api_(Api::createApiForTest(stats_, time_system_)),
+        dispatcher_("test_thread", *api_, time_system_), scope_("memory_release_test.", stats_) {}
+
+  void initialiseAllocatorManager(uint64_t bytes_to_release, float release_interval_s) {
+    const std::string yaml_config = (release_interval_s > 0)
+                                        ? fmt::format(R"EOF(
+  bytes_to_release: {}
+  memory_release_interval: {}s
+)EOF",
+                                                      bytes_to_release, release_interval_s)
+                                        : fmt::format(R"EOF(
+  bytes_to_release: {}
+)EOF",
+                                                      bytes_to_release);
+    const auto proto_config =
+        TestUtility::parseYaml<envoy::config::bootstrap::v3::MemoryAllocatorManager>(yaml_config);
+    allocator_manager_ = std::make_unique<Memory::AllocatorManager>(
+        dispatcher_, Thread::threadFactoryForTest(), scope_, proto_config);
+  }
+
+  void step(const std::chrono::milliseconds& step) {
+    time_system_.advanceTimeAndRun(std::chrono::milliseconds(step), dispatcher_,
+                                   Event::Dispatcher::RunType::NonBlock);
   }
 
-  std::unique_ptr<Memory::Allocator> memory_allocator_;
+  Envoy::Stats::TestUtil::TestStore stats_;
+  Event::SimulatedTimeSystem time_system_;
+  Api::ApiPtr api_;
+  Event::DispatcherImpl dispatcher_;
+  Envoy::Stats::TestUtil::TestScope scope_;
+  std::unique_ptr<Memory::AllocatorManager> allocator_manager_;
 };
 
-TEST_F(MemoryReleaseTest, RelaseRateZeroNoRelease) {
-  auto ptr = std::make_unique<MyStruct>();
-  uint64_t before = Stats::totalCurrentlyAllocated();
-  EXPECT_LOG_NOT_CONTAINS("info",
-                          "Configured tcmalloc with background release rate: 0 bytes per second",
-                          initialiseAllocator(0 /*bytes per second*/));
-  uint64_t after = Stats::totalCurrentlyAllocated();
-  EXPECT_GE(sizeof(MyStruct), after - before);
+TEST_F(MemoryReleaseTest, RelaseRateAboveZeroDefaultIntervalMemoryReleased) {
+  size_t initial_allocated_bytes = Stats::totalCurrentlyAllocated();
+  auto a = std::make_unique<unsigned char[]>(MB);
+  auto b = std::make_unique<unsigned char[]>(MB);
+  if (Stats::totalCurrentlyAllocated() - initial_allocated_bytes <= 0) {
+    // Cannot measure memory usage precisely on this platform.
+    return;
+  }
+  auto initial_unmapped_bytes = Stats::totalPageHeapUnmapped();
+  EXPECT_LOG_CONTAINS(
+      "info", "Configured tcmalloc with background release rate: 512 bytes per 1000 milliseconds",
```

Comment:
do we run into flakes here depending on how the allocation is done as there might not be bytes available for tcmalloc in a releasable form?